### DISCUSSION TITLE
Importing requirements

### DISCUSSION THREAD

**Discussion by mnaderhirn** [dbinfrago/py-capellambse#267] [Q&A]
I have an empty Capella project and I want to import requirements through py-capellambse. 

I have followed the tutorial but was not able to apply the same steps. 
[https://dsd-dbs.github.io/py-capellambse/examples/06%20Introduction%20to%20Requirement%20access%20and%20management.html](https://dsd-dbs.github.io/py-capellambse/examples/06%20Introduction%20to%20Requirement%20access%20and%20management.html)

Since my Capella project is empty i started with creating a requirements_module which appears then in the file

    model.oa.requirement_modules.create(identifier='Requirements', name='Requirements')

Following the tutorial I wanted to create the requirement using

    model.oa.requirement_modules[0].requirements.create(name="New showcase req1") 

which stops my python skript with this comment 

```python
self = <'GenericElement.requirements' ElementRelationAccessor>, elmlist = []
type_hints = ()
kw = {'description': 'description', 'identifier': 'REQ-0001', 'name': 'Requirement', 'related': []}

    def create(
        self,
        elmlist: ElementListCouplingMixin,
        /,
        *type_hints: str | None,
        **kw: t.Any,
    ) -> T:
        """Create and return a new element of type ``elmclass``.
    
        Parameters
        ----------
        elmlist
            The (coupled)
            :py:class:`~capellambse.model.common.element.ElementList` to
            insert the new object into.
        type_hints
            Hints for finding the correct type of element to create. Can
            either be a full or shortened ``xsi:type`` string, or an
            abbreviation defined by the specific Accessor instance.
        kw
            Initialize the properties of the new object. Depending on
            the object's type, some attributes may be required.
        """
>       raise TypeError("Cannot create objects")
E       TypeError: Cannot create objects
```

The following variables are also empty
model.oa.requirements = []
model.oa.requirement_tpyes_folder = []

Since I dont know how to proceed it would be great to get some help. Thank you,
Michael

**Comment by Wuestengecko** [comment]
Hi! Unfortunately I can't reproduce the issue you're having locally, so we'll need to get some more info out of your setup.

First of all: Which version of capellambse are you running – do you have the latest version? You can check this with:

    import capellambse
    print(capellambse.__version__)

If it's not the latest (currently [v0.5.18](https://github.com/DSD-DBS/py-capellambse/releases/tag/v0.5.18)), please update with `pip install -U capellambse` . This might as well already fix it.

---

The error output says: `` kw = {'description': 'description', 'identifier': 'REQ-0001', 'name': 'Requirement', 'related': []} ``

However, in the command above you only copied the command from the demo notebook `` ...create(name="New showcase req1") ``, which should result in `` kw = {'name': 'New showcase req1'} ``. Can you please double-check the exact code you're running there, and make sure that you're actually calling `.requirements.create()` on a `CapellaModule` or `Folder` instance? These are the only places where it's possible to create new requirements, other places can only link to existing ones.

---

If it still doesn't work, we'll have to investigate a bit deeper – in that case, please provide more of the code you're running. If you can, attaching an entire Jupyter notebook which shows the error would be good (you don't necessarily have to provide the model or the source file with your actual requirements).

**Reply by Luffy-cpu** [reply]
Hi,

I am facing a problem similar to this one. I already asked my question in an other discussion, but the title of this discussion looks closer to what I am looking for, so I ask here.
I have a dict in python that contains the new values of requirements attributes, for example
req = {"Id": "R1", "Text": "The requirement shall be complete", "Rationale": "From Specification Doc"}
I assume that in the Capella model, the corresponding attributes are "identifier", "text" and "rationale" (the last one is a custom attribute).
I want to update the text and rationale attributes of the requirement which identifier is equal to "R1" in the Capella model with my dict.
Do I need to call the method .requirements.create() on CapellaModule or Folder? Or is it possible to call another method that just update the existing requirement with the new attributes values?

Thanks in advance,
Christian

**Reply by dwdbmeagle** [reply]
I don't know about updating an existing requirement yet but I have managed to create many new requirements linked to a particular requirement viewpoint folder (ie using a folder to capture the requirements per document) by importing the data from a table in a document.
`new_requirement = model.sa.requirement_modules[1].requirements.create(name=new_req_name,text=new_req_text)`
